livedata/src/main/java/io/getstream/chat/android/livedata/EventBatchUpdate.kt

```diff
@@ -0,0 +1,119 @@
+    val channels: MutableMap<String, ChannelEntity> = mutableMapOf()
+
+    val messages: MutableMap<String, MessageEntity> = mutableMapOf()
+    private val channelsToFetch = mutableSetOf<String>()
+    private val messagesToFetch = mutableSetOf<String>()
+
+    private lateinit var channelMap: Map<String, ChannelEntity>
+    private lateinit var messageMap: Map<String, MessageEntity>
+    private var fetchCompleted: Boolean = false
+
+    fun addMessageData(cid: String, message: Message) {
+        require(fetchCompleted) { "be sure to run batch.fetch before calling this method" }
+        addMessage(MessageEntity(message), message.users())
+
+        getCurrentChannel(cid)?.let {
+            it.updateLastMessage(MessageEntity(message))
+            addChannelEntity(it, emptyList())
+        }
+    }
+
+    fun addChannel(channel: Channel) {
+        // ensure we store all users for this channel
+        addUsers(channel.users())
+        // TODO: this overwrites members which in the case when you have > 100 members isn't the right behaviour
+        channels[channel.cid] = ChannelEntity(channel)
+    }
+
+    fun addChannelEntity(channelEntity: ChannelEntity, channelUsers: List<User>) {
+        // ensure we store all users for this channel
+        addUsers(channelUsers)
+        channels[channelEntity.cid] = channelEntity
+    }
+
+    fun getCurrentChannel(cId: String): ChannelEntity? {
+        require(fetchCompleted) { "be sure to run batch.fetch before calling this method" }
+        return channels[cId] ?: channelMap[cId]
+    }
+
+    fun getCurrentMessage(messageId: String): MessageEntity? {
+        require(fetchCompleted) { "be sure to run batch.fetch before calling this method" }
+        return messages[messageId] ?: messageMap[messageId]
+    }
+
+    fun addMessage(messageEntity: MessageEntity, messageUsers: List<User>) {
+        // ensure we store all users for this channel
+        addUsers(messageUsers)
+        messages[messageEntity.id] = messageEntity
+    }
+
+    fun addUsers(newUsers: List<User>) {
+        users.putAll(newUsers.associateBy(User::id))
+    }
+
+    fun addUser(newUser: User) {
+        addUsers(listOf(newUser))
+    }
+
+    fun addToFetchChannels(cIds: List<String>) {
+        channelsToFetch += cIds
+    }
+
+    fun addToFetchChannels(cId: String) {
+        channelsToFetch += listOf(cId)
+    }
+
+    fun addToFetchMessages(ids: List<String>) {
+        messagesToFetch += ids
+    }
+
+    fun addToFetchMessages(id: String) {
+        messagesToFetch += id
+    }
+
+    suspend fun fetch() {
+        messageMap = domainImpl.repos.messages.select(messagesToFetch.toList()).associateBy { it.id }
+        channelMap = domainImpl.repos.channels.select(channelsToFetch.toList()).associateBy { it.cid }
+        fetchCompleted = true
+    }
+
+    suspend fun execute() {
+        require(fetchCompleted) { "be sure to run batch.fetch before calling this method" }
```

Comment:
In case`fetch()` was already called and then another `add..()` functions called the `fetchCompleted` is already set to `true` allowing to skip another `fetch()` call before calling `execute()`. 

E.g.  
```kotlin 
EventBatchUpdate(domainImpl).apply {
  fetch()
  add...()
  execute()
}
```

This `require()` assertion doesn't sound look safe.